jolt-core/src/main/java/com/bazaarvoice/jolt/templatr/function/Math.java

```diff
@@ -18,151 +18,316 @@
+        }
+        else if(arg instanceof String) {
+            Optional<Number> optional = toNumber( arg );
+            if ( optional.isPresent() ) {
+                return Optional.of( optional.get().longValue() );
+            }
+            else {
                 return Optional.empty();
             }
+        }
+        else {
+            return Optional.empty();
+        }
+    }
 
-            for(Object arg: args) {
-                if(arg instanceof Integer) {
-                    minInt = java.lang.Math.min( minInt, (Integer) arg );
-                }
-                else if(arg instanceof Double) {
-                    minDouble = java.lang.Math.min( minDouble, (Double) arg );
-                }
-                else if(arg instanceof Long) {
-                    minLong = java.lang.Math.min( minLong, (Long) arg );
-                }
+    public static Optional<Double> toDouble(Object arg) {
+        if ( arg instanceof Number ) {
+            return Optional.of( ( (Number) arg ).doubleValue() );
+        }
+        else if(arg instanceof String) {
+            Optional<Number> optional = toNumber( arg );
+            if ( optional.isPresent() ) {
+                return Optional.of( optional.get().doubleValue() );
             }
-            if(minInt == Integer.MAX_VALUE && minDouble == Double.MAX_VALUE) {
+            else {
                 return Optional.empty();
             }
-            // explicit if else to avoid autoboxing
-            if(minInt.longValue() <= minDouble.longValue() && minInt.longValue() <= minLong) {
-                return Optional.<Object>of(minInt);
+        }
+        else {
+            return Optional.empty();
+        }
+    }
+
+    @SuppressWarnings( "unchecked" )
+    public static final class MaxOf implements Function {
+        @Override
+        public Optional<Object> apply( final Object... args ) {
+            if(args.length == 1 && args[0] instanceof List ) {
+                return (Optional) MaxOf( ( (List<Object>) args[0] ).toArray() );
+            }
+            else {
+                return (Optional) MaxOf( args );
             }
-            else if(minLong <= minDouble.longValue() && minLong <= minInt.longValue()) {
-                return Optional.<Object>of(minLong);
+        }
+    }
+
+    @SuppressWarnings( "unchecked" )
+    public static final class MinOf implements Function {
+        @Override
+        public Optional<Object> apply( final Object... args ) {
+            if(args.length == 1 && args[0] instanceof List ) {
+                return (Optional) MinOf( ( (List<Object>) args[0] ).toArray() );
             }
             else {
-                return Optional.<Object>of(minDouble);
+                return (Optional) MinOf( args );
             }
         }
     }
 
-    public static final class Abs implements Function {
+    @SuppressWarnings( "unchecked" )
+    public static final class AbsOf implements Function {
```

Comment:
Abs, toInteger, toLong, toDouble all look like they could just have a base class, that does all the instance of stuff, and calls an protected "toWhatever" method.